src/widget/switch.rs

```diff
@@ -0,0 +1,242 @@
+use crate::piet::{FontBuilder, Text, TextLayout, TextLayoutBuilder};
+use crate::piet::{LinearGradient, RenderContext, UnitPoint};
+use crate::theme;
+use crate::widget::Align;
+use crate::{
+    BaseState, BoxConstraints, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+#[derive(Debug, Clone)]
+pub struct Switch;
+
+impl Switch {
+    pub fn new() -> impl Widget<bool> {
+        Align::vertical(UnitPoint::CENTER, SwitchRaw::default())
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SwitchRaw {
+    knob_pos: Point,
+    knob_hovered: bool,
+}
+
+impl SwitchRaw {
+    fn knob_hit_test(&self, knob_width: f64, mouse_pos: Point) -> bool {
+        let knob_circle = Circle::new(self.knob_pos, knob_width / 2.);
+        if mouse_pos.distance(knob_circle.center) < knob_circle.radius {
+            return true;
+        }
+        false
+    }
+
+    fn paint_label(
+        &mut self,
+        paint_ctx: &mut PaintCtx,
+        base_state: &BaseState,
+        data: &bool,
+        env: &Env,
+        switch_width: f64,
+        switch_padding: f64,
+    ) {
+        let font_name = env.get(theme::FONT_NAME);
+        let font_size = env.get(theme::TEXT_SIZE_NORMAL);
+
+        let label = if *data { "ON" } else { "OFF" };
+
+        let font = paint_ctx
+            .text()
+            .new_font_by_name(font_name, font_size)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let text_layout = paint_ctx
+            .text()
+            .new_text_layout(&font, label)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let mut origin = UnitPoint::LEFT.resolve(Rect::from_origin_size(
+            Point::ORIGIN,
+            Size::new(
+                (base_state.size().width - text_layout.width()).max(0.0),
+                base_state.size().height + (font_size * 1.2) / 2.,
+            ),
+        ));
+
+        // adjust label position
+        origin.y = origin.y.min(base_state.size().height);
+
+        if *data {
+            origin.x = switch_padding * 2.
+        } else {
+            origin.x = switch_width - text_layout.width() - switch_padding * 2.
+        }
+
+        paint_ctx.draw_text(&text_layout, origin, &env.get(theme::LABEL_COLOR));
+    }
+}
+
+impl Widget<bool> for SwitchRaw {
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, base_state: &BaseState, data: &bool, env: &Env) {
+        let switch_padding = 3.;
+        let switch_height = env.get(theme::BORDERED_WIDGET_HEIGHT);
+        let switch_width = switch_height * 2.75;
+        let knob_size = switch_height - 2. * switch_padding;
+
+        let background_rect = RoundedRect::from_origin_size(
+            Point::ORIGIN,
+            Size::new(switch_width, switch_height).to_vec2(),
+            switch_height / 2.,
+        );
+
+        // paint different background for on and off state
+        let background_gradient = if *data {
```

Comment:
I have an orange "caution" color that none of the widgets have used yet. Could potentially use it here for the off state, but this also looks fine as-is. Orange might be too intense for a typical switch use case? [Here's the figma if you're interested](https://www.figma.com/file/YudE0StDEPdemyIIVdUMzr/Druid-GUI-concept?node-id=0%3A1).

Reply:
I think that as future work color should be configurable.